### PR TITLE
NS-CACHE: Support S3 operations for cache enabled buckets

### PR DESCRIPTION
**Pull request by guptsanj**
### Explain the changes
1. Reject any versioning request to buckets with cache property
2. Server Side copy for a cache miss for a source bucket different than the target bucket
3. Multipart upload copy from a same/different sources
4. List objects from cache/hub
5. List Object versions defaults to List objects for cache enabled buckets
6. List multipart uploads from hub
7. Handle add/remove/get object tags from cache/hub buckets
8. Store last modified time from hub for cached objects. Return this value as the last modified time with cached objects.
Update Unit test

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

### CODE REVIEW

File: src/sdk/object_sdk.js [hunk 1]
```diff
@@ -95,6 +95,15 @@ class ObjectSDK {
         return bucket.namespace;
     }
 
+    async read_bucket_sdk_caching_info(name) {
+        try {
+            const { bucket } = await bucket_namespace_cache.get_with_cache({ sdk: this, name });
+            return bucket.namespace.caching;
```
Comment (jeniawhite) [inline src/sdk/object_sdk.js:101]: This crashes if we get here with a regular non-namespace bucket.
Since you are calling that on every request in s3_rest, you will just crash the S3 endpoint and all requests on non-namespaced buckets won't work.

**Review by jeniawhite — DISMISSED** [review]

File: src/endpoint/s3/ops/s3_get_bucket.js [hunk 2]
```diff
@@ -29,7 +29,9 @@ async function get_bucket(req) {
         key_marker: list_type === '2' ?
             (cont_tok_to_key_marker(cont_tok) || start_after) : req.query.marker,
     };
-
+    if (req.query.get_from_cache !== undefined) {
```
[the remark this answers is not in the document]
Comment (guymguym) [reply src/endpoint/s3/ops/s3_get_bucket.js:32]:
```suggestion
    if (req.query.get_from_cache) {
```

File: src/sdk/object_sdk.js (hunk 1 above)
Comment (guymguym) [reply src/sdk/object_sdk.js:101]: fixed

File: src/endpoint/s3/ops/s3_get_bucket.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (guymguym) [reply src/endpoint/s3/ops/s3_get_bucket.js:32]: get_from_cache was added for testing.
Open technical debt with options to implement:
- HTTP `Cache-Control` header
- Custom Header `X-NooBaa-Cache-Control`

File: src/sdk/object_sdk.js [hunk 3]
```diff
@@ -443,9 +452,16 @@ class ObjectSDK {
...
+                params.copy_source.obj_id = source_md.obj_id;
+                params.size = source_params.end - source_params.start;
             }
+
+            source_params.object_md = source_md;
+            source_params.obj_id = source_md.obj_id;
+
             params.source_stream = await source_ns.read_object_stream(source_params, this);
-            if (params.size > (100 * size_utils.MEGABYTE)) {
+            if (params.size && params.size > (100 * size_utils.MEGABYTE)) {
```
Comment (guymguym) [inline src/sdk/object_sdk.js:464]: Can you explain what did you try to achieve with these additions? 
It doesn't look right to me - you use copy_source when not the same namespace?

File: src/endpoint/s3/s3_rest.js [hunk 4]
```diff
@@ -152,6 +152,13 @@ async function handle_request(req, res) {
     usage_report.s3_usage_info.total_calls += 1;
     usage_report.s3_usage_info[op_name] = (usage_report.s3_usage_info[op_name] || 0) + 1;
 
+
+    const caching = await req.object_sdk.read_bucket_sdk_caching_info(req.params.bucket);
+    if (req.query && req.query.VersionId && caching) {
```
[the remark this answers is not in the document]
Comment (guymguym) [reply src/endpoint/s3/s3_rest.js:157]: @guptsanj Can you please check first `if (req.query && req.query.VersionId)` and only then await the caching info? That would remove an unnecessary await in the common handling of any S3 operation so it's good to be efficient here.

[the remark this answers is not in the document]
Comment (guptsanj) [reply src/endpoint/s3/s3_rest.js]: Done.
DeleteObjects version check is done in `namespace_cache.js: delete_multiple_objects`

File: src/sdk/namespace_s3.js [hunk 5]
```diff
@@ -228,11 +245,27 @@ class NamespaceS3 {
...
+                if (response.httpResponse.statusCode > 200) return; // will be handled by error event
+                last_modify_date = response.httpResponse.headers.date;
+                statusCode = response.httpResponse.statusCode;
+                return response;
+            }).promise();
         }
         dbg.log0('NamespaceS3.upload_object:', this.bucket, inspect(params), 'res', inspect(res));
         const etag = s3_utils.parse_etag(res.ETag);
-        return { etag, version_id: res.VersionId };
+        return { etag, statusCode: statusCode, version_id: res.VersionId, date: last_modify_date };
```
[the remark this answers is not in the document]
Comment (guptsanj) [reply src/sdk/namespace_s3.js]: I was using the statusCode to check if the upload operation succeeded. Instead of using `statusCode`, I just check if the date is set in the response in namespace_cache.js

File: src/endpoint/s3/ops/s3_get_bucket.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (guptsanj) [reply src/endpoint/s3/ops/s3_get_bucket.js]: This will be handled in a new issue. Leaving this unresolved until the issue is created.

File: src/sdk/object_sdk.js (hunk 3 above)
Comment (guptsanj) [reply src/sdk/object_sdk.js]:
```
                params.copy_source.bucket = actual_source_ns.get_bucket(bucket);
                params.copy_source.obj_id = source_md.obj_id;
                params.size = source_params.end - source_params.start;
```
This was added for a case where a byte range is copied from an object bucket into a namespace bucket.
The test is at: https://github.com/noobaa/noobaa-core/blob/684f4e672eda381e9f63159c51d3c5a9cf61af79/src/test/unit_tests/test_s3_ops.js#L384

Comment (guptsanj) [reply src/sdk/object_sdk.js]: @guymguym 
```
            source_params.object_md = source_md;
            source_params.obj_id = source_md.obj_id;
```
Added because of exceptions thrown when 
https://github.com/noobaa/noobaa-core/blob/684f4e672eda381e9f63159c51d3c5a9cf61af79/src/test/unit_tests/test_s3_ops.js#L344
is executed.

Comment (guptsanj) [reply src/sdk/object_sdk.js]: The tests which give this exception are
https://github.com/noobaa/noobaa-core/blob/684f4e672eda381e9f63159c51d3c5a9cf61af79/src/test/unit_tests/test_s3_ops.js#L509
https://github.com/noobaa/noobaa-core/blob/684f4e672eda381e9f63159c51d3c5a9cf61af79/src/test/unit_tests/test_s3_ops.js#L513

File: src/endpoint/s3/s3_utils.js [hunk 6]
```diff
@@ -549,6 +554,13 @@ function parse_website_to_body(website) {
     return reply;
 }
 
+function get_http_response_from_ret(res) {
+    const r = res.$response;
+    if (!r) throw new Error("no $response in s3 returned object");
+    if (!r.httpResponse.headers.date) throw new Error("...");
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:560]: Edit the error text

File: src/sdk/namespace_cache.js [hunk 7]
```diff
@@ -157,18 +168,27 @@ class NamespaceCache {
...
+            });
+        }
+        return this.namespace_hub.list_objects(params, object_sdk);
     }
 
     async list_uploads(params, object_sdk) {
-        // TODO listing from cache only for deevelopment
-        return this.namespace_nb.list_uploads(params, object_sdk);
+
+        return this.namespace_hub.list_uploads(params, object_sdk);
```
Comment (guymguym) [inline src/sdk/namespace_cache.js:186]: How about we use the same testing flag `req.query.get_from_cache` to allow both modes?
CC @paulxuzhang

File: src/sdk/namespace_cache.js [hunk 8]
```diff
@@ -157,18 +168,27 @@ class NamespaceCache {
...
     }
 
     async list_object_versions(params, object_sdk) {
-        // TODO listing from cache only for deevelopment
-        return this.namespace_nb.list_object_versions(params, object_sdk);
+        // Always return latest version.
+        return this.list_objects(params, object_sdk);
```
Comment (guymguym) [inline src/sdk/namespace_cache.js:191]: So this means that even if we have versions in the hub we do not list those.
I don't think this is what we should be doing.
I think we should list version from hub, and just avoid caching versions other than latest on GET.

File: src/test/unit_tests/test_namespace_cache.js [hunk 9]
```diff
@@ -239,26 +239,27 @@ class MockNamespace {
 
                 this._buf = Buffer.concat(recv_buf);
                 const etag = this._md5.digest('hex');
-
+                const create_time = new Date().getTime();
```
Comment (guymguym) [inline src/test/unit_tests/test_namespace_cache.js:242]:
```suggestion
                const create_time = Date.now();
```

File: src/sdk/namespace_s3.js [hunk 10]
```diff
@@ -253,7 +253,8 @@ class NamespaceS3 {
         }
         dbg.log0('NamespaceS3.upload_object:', this.bucket, inspect(params), 'res', inspect(res));
         const etag = s3_utils.parse_etag(res.ETag);
-        return { etag, version_id: res.VersionId };
+        const last_modified_time = s3_utils.get_http_response_from_ret(res).httpResponse.headers.date;
+        return { etag, version_id: res.VersionId, date: last_modified_time };
```
Comment (guymguym) [inline src/sdk/namespace_s3.js:257]: why not return it as `last_modified_time`?

File: src/sdk/namespace_s3.js [hunk 11]
```diff
@@ -253,7 +253,8 @@ class NamespaceS3 {
         }
         dbg.log0('NamespaceS3.upload_object:', this.bucket, inspect(params), 'res', inspect(res));
         const etag = s3_utils.parse_etag(res.ETag);
-        return { etag, version_id: res.VersionId };
+        const last_modified_time = s3_utils.get_http_response_from_ret(res).httpResponse.headers.date;
```
Comment (guymguym) [inline src/sdk/namespace_s3.js:256]: I suggest this will be it's own utility function `s3_utils.get_http_response_date(req)` that uses `s3_utils.get_http_response_from_ret()` inside, but it's best to hide these details from the namespace code if possible.

**Review by guymguym — CHANGES_REQUESTED** [review]

File: src/sdk/namespace_cache.js (hunk 8 above)
Comment (guptsanj) [reply src/sdk/namespace_cache.js:196]: The `list_objects` function will return from the hub or from the cache based on the `get_from_cache` param value. So this should return from the hub if there are versions on the hub. I have updated the comment

File: src/endpoint/s3/s3_utils.js [hunk 12]
```diff
@@ -573,3 +590,5 @@ exports.parse_lock_header = parse_lock_header;
 exports.parse_body_object_lock_conf_xml = parse_body_object_lock_conf_xml;
 exports.parse_to_camel_case = parse_to_camel_case;
 exports._is_valid_retention = _is_valid_retention;
+exports.get_http_response_from_ret = get_http_response_from_resp;
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:593]: please fix the exported name and function name to match

File: src/sdk/namespace_s3.js [hunk 13]
```diff
@@ -253,7 +253,8 @@ class NamespaceS3 {
         }
         dbg.log0('NamespaceS3.upload_object:', this.bucket, inspect(params), 'res', inspect(res));
         const etag = s3_utils.parse_etag(res.ETag);
-        return { etag, version_id: res.VersionId };
+        const last_modified_time = s3_utils.get_http_response_date(res);
+        return { etag, version_id: res.VersionId, last_modified_time: last_modified_time };
```
Comment (guymguym) [inline src/sdk/namespace_s3.js:257]:
```suggestion
        return { etag, version_id: res.VersionId, last_modified_time };
```

File: src/sdk/namespace_cache.js [hunk 14]
```diff
@@ -157,18 +168,32 @@ class NamespaceCache {
     /////////////////
 
     async list_objects(params, object_sdk) {
-        // TODO listing from cache only for deevelopment
-        return this.namespace_nb.list_objects(params, object_sdk);
+        const get_from_cache = params.get_from_cache;
+        params = _.omit(params, 'get_from_cache');
+        if (get_from_cache) {
+            return this.namespace_nb.list_objects(params, object_sdk).then(res => {
```
Comment (guymguym) [inline src/sdk/namespace_cache.js:174]: Avoid using `promise.then()`, use `await` instead -
```suggestion
            const res = await this.namespace_nb.list_objects(params, object_sdk);
```

File: src/sdk/namespace_cache.js [hunk 15]
```diff
@@ -157,18 +168,32 @@ class NamespaceCache {
...
+                _.forEach(res.objects, obj => {
+                    obj.create_time = obj.last_modified_time ? obj.last_modified_time : obj.create_time;
```
Comment (guymguym) [inline src/sdk/namespace_cache.js:176]: Why is this needed? Shouldn't that be in our s3 level like we do in `s3_utils.set_response_object_md()`?

File: src/endpoint/s3/ops/s3_get_bucket.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (guymguym) [reply src/endpoint/s3/ops/s3_get_bucket.js]: opened https://github.com/noobaa/noobaa-core/issues/6147. resolving.

File: src/server/object_services/schemas/object_md_schema.js [hunk 16]
```diff
@@ -77,7 +77,7 @@ module.exports = {
         upload_started: { objectid: true },
         create_time: { date: true },
         cache_last_valid_time: { date: true },
-
+        last_modified_time: { date: true },
```
Comment (guymguym) [inline src/server/object_services/schemas/object_md_schema.js:80]: Can you please add a comment that explains this new meta-data property.
Also I think a comment for previous line's cache_last_valid_time will be good.
```suggestion
        // last_modified_time is an optional property that is set on cache buckets
        // to separate the times of creation of the cache object (create_time)
        // vs the "real" LastModifiedTime from the hub.
        last_modified_time: { date: true },
```

File: src/sdk/namespace_cache.js (hunk 15 above)
Comment (guptsanj) [reply src/sdk/namespace_cache.js:176]: The object listing does not go through `s3_utils.set_response_object_md()`

Comment (guymguym) [reply src/sdk/namespace_cache.js:176]: ok, but it seems that this handling is needed in s3_get_bucket instead of just here in namespace_cache.

File: src/sdk/namespace_cache.js [hunk 17]
```diff
@@ -486,9 +511,19 @@ class NamespaceCache {
...
+                }
+                const update_params = _.pick(_.defaults(
+                    { bucket: this.namespace_nb.target_bucket }, params), 'bucket', 'key', 'last_modified_time');
+                update_params.last_modified_time = (new Date(res.last_modified_time)).getTime();
+                setImmediate(() => object_sdk.rpc_client.object.update_object_md(update_params));
+                return res;
+            };
             // defer the final callback of the cache stream until the hub ack
-            const cache_finalizer = callback => hub_promise.then(() => callback(), err => callback(err));
+            const cache_finalizer = hub_promise.then(data => hub_callback(data), err => hub_callback(err));
```
Comment (guymguym) [inline src/sdk/namespace_cache.js:526]: The type of `cache_finalizer`  should be a function and you assign it a promise - that won't work.
What we wanted to achieve was to hold the cache stream open until the hub promise completes.
So the previous code did that by setting cache_finalizer to a function that puts the cache stream final callback as the callback of the hub_promise.

Comment (guymguym) [reply src/sdk/namespace_cache.js:526]:
```
function stream_final_from_promise(promise) {
   return callback => promise.then(() => callback(), err => callback(err));
}

const cache_final_promise = hub_promise.then(async upload_res => {
   await object_sdk.rpc_client.object.update_object_md(update_params);
});
const cache_finalizer = stream_final_from_promise(cache_final_promise);

```

File: src/sdk/object_sdk.js (hunk 3 above)
Comment (guymguym) [reply src/sdk/object_sdk.js]: so we want to remove these changes and see why the test is failing,
because it might be a new test case that is the cause

File: src/sdk/namespace_cache.js (hunk 8 above)
Comment (guymguym) [reply src/sdk/namespace_cache.js:196]: ok

File: src/endpoint/s3/ops/s3_get_bucket.js [hunk 18]
```diff
@@ -30,7 +30,18 @@ async function get_bucket(req) {
...
 
+    if (req.query.get_from_cache) {
+        params.get_from_cache = req.query.get_from_cache;
+    }
     const reply = await req.object_sdk.list_objects(params);
+
+    // Update the object creation time in case this is from cache
+    if (req.query.get_from_cache) {
+        _.forEach(reply.objects, obj => {
+            obj.create_time = obj.last_modified_time ? obj.last_modified_time : obj.create_time;
```
Comment (guymguym) [inline src/endpoint/s3/ops/s3_get_bucket.js:41]: You know, we don't really want to modify these objects properties meaning,
why not put this inside the response building in line 69 like so:
```
                    // if the object specifies last_modified_time we use it, otherwise take create_time.
                    LastModified: s3_utils.format_s3_xml_date(obj.last_modified_time || obj.create_time),
```

File: src/endpoint/s3/ops/s3_get_bucket.js [hunk 19]
```diff
@@ -30,7 +30,18 @@ async function get_bucket(req) {
             (cont_tok_to_key_marker(cont_tok) || start_after) : req.query.marker,
     };
 
+    if (req.query.get_from_cache) {
+        params.get_from_cache = req.query.get_from_cache;
+    }
     const reply = await req.object_sdk.list_objects(params);
+
+    // Update the object creation time in case this is from cache
+    if (req.query.get_from_cache) {
```
Comment (guymguym) [inline src/endpoint/s3/ops/s3_get_bucket.js:39]: I don't think this is the condition is needed, I think obj.last_modified_time should always have precedence over create_time if the former is set.

**Review by guymguym — APPROVED** [review]
Looks awesome!
